File: mobile-app-nexLevel/src/utils/phoneValidation.js
/**
 * التحقق من صحة أرقام الهواتف حسب الدولة.
 * يطابق قواعد الباك اند - يرفض 0 في البداية ويطابق طول الرقم لكل دولة.
 *
 * Rules: dialCode => { min, max, errorKey }
 * errorKey يستخدم مع i18n: phoneValidation.{errorKey}
 */
const RULES = {
  '964': { min: 10, max: 10, errorKey: 'iq' },
  '966': { min: 9, max: 9, errorKey: 'sa' },
  '20': { min: 9, max: 10, errorKey: 'eg' },
  '971': { min: 9, max: 9, errorKey: 'ae' },
  '962': { min: 9, max: 9, errorKey: 'jo' },
  '965': { min: 8, max: 8, errorKey: 'kw' },
  '974': { min: 8, max: 8, errorKey: 'qa' },
  '973': { min: 8, max: 8, errorKey: 'bh' },
  '968': { min: 8, max: 8, errorKey: 'om' },
  '967': { min: 9, max: 9, errorKey: 'ye' },
  '963': { min: 9, max: 9, errorKey: 'sy' },
  '961': { min: 8, max: 8, errorKey: 'lb' },
  '970': { min: 9, max: 9, errorKey: 'ps' },
  '218': { min: 9, max: 9, errorKey: 'ly' },
  '216': { min: 8, max: 8, errorKey: 'tn' },
  '213': { min: 9, max: 9, errorKey: 'dz' },
  '212': { min: 9, max: 9, errorKey: 'ma' },
  '249': { min: 9, max: 9, errorKey: 'sd' },
  '1': { min: 10, max: 10, errorKey: 'us' },
  '44': { min: 10, max: 10, errorKey: 'gb' },
  '33': { min: 9, max: 9, errorKey: 'fr' },
  '49': { min: 10, max: 11, errorKey: 'de' },
  '39': { min: 9, max: 11, errorKey: 'it' },
  '90': { min: 10, max: 10, errorKey: 'tr' },
  '91': { min: 10, max: 10, errorKey: 'in' },
  '92': { min: 10, max: 10, errorKey: 'pk' },
  '98': { min: 10, max: 10, errorKey: 'ir' },
  '86': { min: 11, max: 11, errorKey: 'cn' },
  '81': { min: 10, max: 10, errorKey: 'jp' },
  '7': { min: 10, max: 10, errorKey: 'ru' },
  '62': { min: 9, max: 12, errorKey: 'id' },
  '234': { min: 10, max: 10, errorKey: 'ng' },
  '27': { min: 9, max: 9, errorKey: 'za' },
  '254': { min: 9, max: 9, errorKey: 'ke' },
  '251': { min: 9, max: 9, errorKey: 'et' },
  '880': { min: 10, max: 10, errorKey: 'bd' },
  '93': { min: 9, max: 9, errorKey: 'af' }
}

/**
 * إزالة الصفر من بداية الرقم
 */
export function normalizeNationalNumber(phone) {
  let digits = (phone || '').replace(/\D/g, '')
  while (digits.length > 1 && digits[0] === '0') {
    digits = digits.slice(1)
  }
  return digits
}

/**
 * التحقق من صحة الرقم
 * @param {string} countryCode - مفتاح الدولة (مثل 964، 966) بدون +
 * @param {string} phone - الرقم كما أدخله المستخدم
 * @param {Function} t - دالة الترجمة useI18n().t
 * @returns {{ valid: boolean, errorKey?: string, normalized?: string }}
 */
export function validatePhone(countryCode, phone, t) {
  const code = (countryCode || '').trim().replace(/^\+\s*/, '').replace(/\s/g, '')
  if (!code || code.length > 4) {
    return { valid: false, errorKey: 'invalidCountryCode' }
  }

  const digits = (phone || '').replace(/\D/g, '')
  if (digits.length > 1 && digits.startsWith('0')) {
    return { valid: false, errorKey: 'noLeadingZero' }
  }

  const national = normalizeNationalNumber(phone || '')
  if (!national) {
    return { valid: false, errorKey: 'required' }
  }

  if (!/^\d+$/.test(national)) {
    return { valid: false, errorKey: 'digitsOnly' }
  }

  const fullPhone = code + national
  if (fullPhone.length > 20) {
    return { valid: false, errorKey: 'tooLong' }
  }

  const rule = RULES[code]
  if (rule) {
    if (national.length < rule.min || national.length > rule.max) {
      return { valid: false, errorKey: rule.errorKey }
    }
  } else {
    if (national.length < 7 || national.length > 15) {
      return { valid: false, errorKey: 'generic' }
    }
  }

  return { valid: true, normalized: national }
}

/**
 * الحصول على رسالة الخطأ المترجمة
 */
export function getPhoneErrorMessage(result, t) {
  if (!result.errorKey) return ''
  const key = `phoneValidation.${result.errorKey}`
  const msg = t(key)
  return msg !== key ? msg : result.errorKey
}

File: flutter-app/lib/core/phone_validation.dart
import 'i18n/i18n.dart';

/// Port of utils/phoneValidation.js.
class PhoneResult {
  const PhoneResult.valid(this.normalized)
      : valid = true,
        errorKey = null;
  const PhoneResult.invalid(this.errorKey)
      : valid = false,
        normalized = null;

  final bool valid;
  final String? errorKey;
  final String? normalized;

  String get message {
    if (errorKey == null) return '';
    final key = 'phoneValidation.$errorKey';
    final msg = t(key);
    return msg != key ? msg : errorKey!;
  }
}

const _rules = <String, (int, int, String)>{
  '964': (10, 10, 'iq'), '966': (9, 9, 'sa'), '20': (9, 10, 'eg'), '971': (9, 9, 'ae'),
  '962': (9, 9, 'jo'), '965': (8, 8, 'kw'), '974': (8, 8, 'qa'), '973': (8, 8, 'bh'),
  '968': (8, 8, 'om'), '967': (9, 9, 'ye'), '963': (9, 9, 'sy'), '961': (8, 8, 'lb'),
  '970': (9, 9, 'ps'), '218': (9, 9, 'ly'), '216': (8, 8, 'tn'), '213': (9, 9, 'dz'),
  '212': (9, 9, 'ma'), '249': (9, 9, 'sd'), '1': (10, 10, 'us'), '44': (10, 10, 'gb'),
  '33': (9, 9, 'fr'), '49': (10, 11, 'de'), '39': (9, 11, 'it'), '90': (10, 10, 'tr'),
  '91': (10, 10, 'in'), '92': (10, 10, 'pk'), '98': (10, 10, 'ir'), '86': (11, 11, 'cn'),
  '81': (10, 10, 'jp'), '7': (10, 10, 'ru'), '62': (9, 12, 'id'), '234': (10, 10, 'ng'),
  '27': (9, 9, 'za'), '254': (9, 9, 'ke'), '251': (9, 9, 'et'), '880': (10, 10, 'bd'),
  '93': (9, 9, 'af'),
};

String normalizeNationalNumber(String phone) {
  var digits = phone.replaceAll(RegExp(r'\D'), '');
  while (digits.length > 1 && digits.startsWith('0')) {
    digits = digits.substring(1);
  }
  return digits;
}

PhoneResult validatePhone(String countryCode, String phone) {
  final code = countryCode.trim().replaceFirst(RegExp(r'^\+\s*'), '').replaceAll(RegExp(r'\s'), '');
  if (code.isEmpty || code.length > 4) return const PhoneResult.invalid('invalidCountryCode');
  final digits = phone.replaceAll(RegExp(r'\D'), '');
  if (digits.length > 1 && digits.startsWith('0')) return const PhoneResult.invalid('noLeadingZero');
  final national = normalizeNationalNumber(phone);
  if (national.isEmpty) return const PhoneResult.invalid('required');
  if (!RegExp(r'^\d+$').hasMatch(national)) return const PhoneResult.invalid('digitsOnly');
  if ((code + national).length > 20) return const PhoneResult.invalid('tooLong');
  final rule = _rules[code];
  if (rule != null) {
    if (national.length < rule.$1 || national.length > rule.$2) return PhoneResult.invalid(rule.$3);
  } else if (national.length < 7 || national.length > 15) {
    return const PhoneResult.invalid('generic');
  }
  return PhoneResult.valid(national);
}
